src/FindIntrinsics.cpp

```diff
@@ -897,6 +897,129 @@ class FindIntrinsics : public IRMutator {
         }
         return op;
     }
+
+    // Narrow and comparisons between ramps and broadcasts to produce masks that
+    // match the bit-width of the type being selected between or loaded or
+    // stored. We do this late in lowering in this pass instead of in the
+    // simplifier because it messes up the reasoning done by loop partitioning.
+    //
+    // For example if we're selecting between uint8s, and we have the condition:
+    // ramp(x, 1, 16) < broadcast(y)
+    // where x is an Int(32), we can rewrite this to:
+    // ramp(0, 1, 16) < broadcast(y - x)
+    // which can be safely narrowed to:
+    // cast<int8_t>(ramp(0, 1, 16)) < saturating_cast<int8_t>(broadcast(y - x))
+    Expr narrow_predicate(const Expr &p, Type t) {
+        if (t.bits() >= 32) {
+            return p;
+        }
+
+        int lanes = t.lanes();
+
+        if (const Or *op = p.as<Or>()) {
+            return narrow_predicate(op->a, t) || narrow_predicate(op->b, t);
+        } else if (const And *op = p.as<And>()) {
+            return narrow_predicate(op->a, t) && narrow_predicate(op->b, t);
+        } else if (const Not *op = p.as<Not>()) {
+            return !narrow_predicate(op->a, t);
+        }
+
+        const LT *lt = p.as<LT>();
+        const LE *le = p.as<LE>();
+        // Check it's a comparison
+        if (!(le || lt)) {
+            return p;
+        }
+        // Check it's an int32 comparison
+        if ((lt ? lt->a.type() : le->a.type()) != Int(32, lanes)) {
+            return p;
+        }
+
+        auto rewrite = IRMatcher::rewriter(p, Int(32, lanes));
+
+        // Construct predicates which state the ramp can't hit the extreme
+        // values of an int8 or an int16. This is an overconservative condition,
```

Comment:
At first I was going to suggest saturating_narrow on the ramp because it's faster (on x86 at least, and possible others?) before I realized the rewritten ramps below are all constant-folded. If we instead used bounds inference to prove this for symbolic ramps, we should probably saturating_cast the ramp as well, right?